derivative/dglobal.py

```diff
@@ -7,40 +7,47 @@
 from scipy import interpolate, sparse
 from scipy.special import legendre
 from sklearn.linear_model import Lasso
+from specderiv import cheb_deriv, fourier_deriv
 
 
 @register("spectral")
 class Spectral(Derivative):
-    def __init__(self, **kwargs):
+    def __init__(self, order=1, axis=0, basis='fourier', filter=None):
         """
-        Compute the numerical derivative by first computing the FFT. In Fourier space, derivatives are multiplication
-        by i*phase; compute the IFFT after.
-
-        Args:
-            **kwargs: Optional keyword arguments.
-
+        Compute the numerical derivative by spectral methods. In Fourier space, derivatives are multiplication
+        by i*phase; compute the inverse transform after. Use either Fourier modes of Chebyshev polynomials as
+        the basis.
+        
         Keyword Args:
-            filter: Optional. A function that takes in frequencies and outputs weights to scale the coefficient at
-                the input frequency in Fourier space. Input frequencies are the discrete fourier transform sample
-                frequencies associated with the domain variable. Look into python signal processing resources in
-                scipy.signal for common filters.
-
+            order (int): order of the derivative, defaults to 1st order
+            axis (int): the dimension of the data along which to differentiate, defaults to first dimension
+            basis (str): 'fourier' or 'chebyshev', the set of basis functions to use for differentiation
+                Note `basis='fourier'` assumes your function is periodic and sampled over a period of its domain,
+                [a, b), and `basis='chebyshev'` assumes your function is sampled at cosine-spaced points on the
+                domain [a, b].
+            filter: Optional. A function that takes in wavenumbers and outputs weights to scale the
+                corresponding mode in frequency space. Input wavenumbers are the k used in the discrete fourier
+                transform. Look into python signal processing resources in scipy.signal for common filters.
         """
-        # Filter function. Default: Identity filter
-        self.filter = kwargs.get('filter', np.vectorize(lambda f: 1))
-        self._x_hat = None
-        self._freq = None
-
-    def _dglobal(self, t, x):
-        self._x_hat = np.fft.fft(x)
-        self._freq = np.fft.fftfreq(t.size, d=(t[1] - t[0]))
+        self.order = order
+        self.axis = axis
+        if basis not in ['chebyshev', 'fourier']:
+            raise ValueError("Only chebyshev and fourier bases are allowed.")
+        self.basis = basis
+        self.filter = filter
+
+    @_memoize_arrays(1) # the memoization is 1 deep, as in only remembers the most recent args
+    def _global(self, t, x):
+        if self.basis == 'chebyshev':
+            return cheb_deriv(x, t, self.order, self.axis, self.filter)
```

Comment:
It struck me just now that filtering by frequency on cosine-spaced points isn’t totally sound: high-frequency variations look lower frequency around the edges where sample density is higher.

Reply:
I've been pondering this more https://github.com/pavelkomarov/spectral-derivatives/issues/14. Basically what I'm currently doing is keeping `filter(k)` amount of the $k{th}$ *Chebyshev polynomial*, versus in the Fourier case keeping `filter(k)` amount of the $k{th}$ *wavenumber*. This may be the best thing to do, most spiritually-aligned with "spectral". I so far haven't thought of a better alternative. But it's counter-intuitive: Because the Chebyshev polynomials have steeper slope at the edges of the domain, we can fit high-frequencies better there. So if the assumption is that noise is high frequency and needs to be separated from the data this way, we can't filter it as consistently using this basis.

Reply:
I've made [a notebook](https://github.com/pavelkomarov/spectral-derivatives/blob/main/notebooks/filtering_noise.ipynb) to empirically compare spectral filtering in the Fourier and Chebyshev bases, and pulled in a little of Floris' PyNumDiff to compare with Savitzky-Golay filtering + Gaussian smoothing. Fourier spectral is the gold standard, but even it can be fragile. Chebyshev can work okay (not *great* but really not worse than `savgoldiff`) for 1st and 2nd derivatives in the presence of a little noise but starts to blow up beyond that.